robot/impl/resource_manager.go

```diff
@@ -537,115 +539,150 @@ func (manager *resourceManager) completeConfig(
+								"model", conf.Model)
+							return
+						}
+					}
 
-			switch {
-			case resName.API.IsComponent(), resName.API.IsService():
-				newRes, newlyBuilt, err := manager.processResource(ctxWithTimeout, conf, gNode, lr)
-				if newlyBuilt || err != nil {
-					if err := manager.markChildrenForUpdate(resName); err != nil {
-						manager.logger.CErrorw(ctx,
-							"failed to mark children of resource for update",
-							"resource", resName,
-							"reason", err)
+					switch {
+					case resName.API.IsComponent(), resName.API.IsService():
+
+						newRes, newlyBuilt, err := manager.processResource(ctxWithTimeout, conf, gNode, lr)
+						if newlyBuilt || err != nil {
+							if err := manager.markChildrenForUpdate(resName); err != nil {
+								manager.logger.CErrorw(ctx,
+									"failed to mark children of resource for update",
+									"resource", resName,
+									"reason", err)
+							}
+						}
+
+						if err != nil {
+							gNode.LogAndSetLastError(
+								fmt.Errorf("resource build error: %w", err),
+								"resource", conf.ResourceName(),
+								"model", conf.Model)
+							return
+						}
+
+						// if the ctxWithTimeout fails with DeadlineExceeded, then that means that
+						// resource generation is running async, and we don't currently have good
+						// validation around how this might affect the resource graph. So, we avoid
+						// updating the graph to be safe.
+						if errors.Is(ctxWithTimeout.Err(), context.DeadlineExceeded) {
+							manager.logger.CErrorw(
+								ctx, "error building resource", "resource", conf.ResourceName(), "model", conf.Model, "error", ctxWithTimeout.Err())
+						} else {
+							gNode.SwapResource(newRes, conf.Model)
+						}
+
+					default:
+						err := errors.New("config is not for a component or service")
+						gNode.LogAndSetLastError(err, "resource", resName)
 					}
-				}
+				})
 
-				if err != nil {
-					gNode.LogAndSetLastError(
-						fmt.Errorf("resource build error: %w", err),
-						"resource", conf.ResourceName(),
-						"model", conf.Model)
-					return
+				select {
+				case <-resChan:
+				case <-ctxWithTimeout.Done():
+					if errors.Is(ctxWithTimeout.Err(), context.DeadlineExceeded) {
+						lr.logger.CWarn(ctx, rutils.NewBuildTimeoutError(resName.String()))
+					}
+				case <-ctx.Done():
+					return ctx.Err()
 				}
+				return nil
+			}
 
-				// if the ctxWithTimeout fails with DeadlineExceeded, then that means that
-				// resource generation is running async, and we don't currently have good
-				// validation around how this might affect the resource graph. So, we avoid
-				// updating the graph to be safe.
-				if errors.Is(ctxWithTimeout.Err(), context.DeadlineExceeded) {
-					manager.logger.CErrorw(
-						ctx, "error building resource", "resource", conf.ResourceName(), "model", conf.Model, "error", ctxWithTimeout.Err())
-				} else {
-					gNode.SwapResource(newRes, conf.Model)
+			var forceSyncRes bool
```

Comment:
Should this be initialized to `forceSync`? It's not clear to me we're obeying that input variable.

Reply:
yeah that would be clearer, will change

Reply:
Just to confirm, this is a bug, yes? If code was calling `reconfigure(forceSync=true)`, this `forceSyncRes` would be false and consequently we'd execute `levelErrG.Go`?

Reply:
ah shoot, yeah this is totally a bug... not just a matter of clarity

Reply:
ugh, and this would only have manifested as a flaky test - the worst!